### PR TITLE
add the attention mask for flash-attention.

### PR DESCRIPTION
**Pull request by MayDomine**

### CODE REVIEW

**Comment by MayDomine** [comment]
> Could we use (-inf, 0) bias to mock mask feature?

Yes, that would be another way to implement mask. But if the input mask has a whole row of false, it will be quite tricky to handle the nan value in the softmax(qk). I wrote this because I have this kind of need that bias can not satisfy.

**Comment by Kite0011** [comment]
i'm curious why this branch not be merged final?

**Comment by Zyriix** [comment]
it this work? why it was closed?

**Comment by nofreewill42** [comment]
"But if the input mask has a whole row of false, it will be quite tricky to handle the nan value in the softmax(qk)."

What about handling it with an assertion?
